src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -0,0 +1,276 @@
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.log;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.console.AnnotatedLargeText;
+import hudson.console.ConsoleAnnotationOutputStream;
+import hudson.model.BuildListener;
+import hudson.model.StreamBuildListener;
+import hudson.model.TaskListener;
+import hudson.util.StreamTaskListener;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.RandomAccessFile;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.logging.Logger;
+import org.apache.commons.io.input.NullReader;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.Beta;
+import org.kohsuke.stapler.framework.io.ByteBuffer;
+
+/**
+ * Simple implementation of log storage in a single file that maintains a side file with an index indicating where node transitions occur.
+ * Each line in the index file is a byte offset, optionally followed by a space and then a node ID.
+ */
+@Restricted(Beta.class)
+public final class FileLogStorage implements LogStorage {
+
+    private static final Logger LOGGER = Logger.getLogger(FileLogStorage.class.getName());
+
+    private static final Map<File, FileLogStorage> openStorages = Collections.synchronizedMap(new HashMap<>());
+
+    public static synchronized LogStorage forFile(File log) {
+        return openStorages.computeIfAbsent(log, FileLogStorage::new);
+    }
+
+    private final File log;
+    private final File index;
+    @SuppressFBWarnings(value = "IS2_INCONSISTENT_SYNC", justification = "FB apparently gets confused by what the lock is, and anyway we only care about synchronizing writes")
+    private FileOutputStream os;
+    private Writer indexOs;
+    private String lastId;
+
+    private FileLogStorage(File log) {
+        this.log = log;
+        this.index = new File(log + "-index");
+    }
+
+    private synchronized void open() throws IOException {
+        if (os == null) {
+            os = new FileOutputStream(log, true);
+            if (index.isFile()) {
+                try (BufferedReader r = Files.newBufferedReader(index.toPath(), StandardCharsets.UTF_8)) {
+                    String lastLine = null;
+                    while (true) {
+                        String line = r.readLine();
+                        if (line == null) {
+                            break;
+                        } else {
+                            lastLine = line;
+                        }
+                    }
+                    if (lastLine != null) {
```

Comment:
What if the content did not finish writing and the "lastId" has been truncated?  (i.e. the line has only partially written)

Reply:
Then the line → step mapping for this build will not be perfect. For example, in classic UI it would mean that some text would be incorrectly not associated with a step block but just shown as part of general output. Too minor to worry about.

Reply:
Sounds like a bug to me, but I'll let you fix it when the first user reports it.